### PR TITLE
Code updated for PR

### PR DESCRIPTION
**Pull request by Kashif-Rehman**
- Api.cy.js file updated
- test.cy.js file updated

### CODE REVIEW

File: cypress/e2e/test.cy.js [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  });
+
+  beforeEach(() => {
+    cy.visit(url);
+    cy.viewport(1920, 1080);
+  });
+
+  it("Verify user can enter new data into table", () => {
+    cy.get(".category-cards > :nth-child(1)").click();
+    cy.get(":nth-child(1) > .element-list > .menu-list > #item-3").click();
```
Comment (nedo) [inline cypress/e2e/test.cy.js:18]: selector is week here, please select strong selector

File: cypress/e2e/test.cy.js [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    cy.get(".rt-tr-group:eq(1) .rt-td:eq(0)").should("contain", "Aldren");
+    cy.get(".rt-tr-group:eq(1) .rt-td:eq(1)").should("contain", "BV");
+  });
+
+  it("verify broken images", () => {
+    cy.get(".category-cards > :nth-child(1)").click();
+    cy.get(":nth-child(1) > .element-list > .menu-list > #item-6").click();
+
+    cy.get('[src="/images/Toolsqa_1.jpg"]').each(($img) => {
```
Comment (nedo) [inline cypress/e2e/test.cy.js:56]: Can You please explain this line of code?

File: cypress/e2e/API.cy.js [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      requestBody = data;
+    });
+  });
+
+  it("Creation of user account", () => {
+    cy.request({
+      method: "POST",
+      url: "https://demoqa.com/Account/v1/User",
+      body: {
+        userName: requestBody.userName,
```
Comment (nedo) [inline cypress/e2e/API.cy.js:18]: From where 'userName' coming ?

File: cypress/e2e/API.cy.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      expect(res.body.username).to.equal("go");
+    });
+  });
+
+  it.only("Add a list of books", () => {
+    cy.request({
+      method: "POST",
+      url: "https://demoqa.com/BookStore/v1/Books",
+      headers: {
+        Authorization: requestBody.auth,
```
Comment (nedo) [inline cypress/e2e/API.cy.js:32]: what is `auth` here ?

File: cypress/e2e/test.cy.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
+///<reference types="cypress"/>
+import "@4tw/cypress-drag-drop";
+describe("Assignment spec", () => {
+  const url = "https://demoqa.com";
```
Comment (nedo) [inline cypress/e2e/test.cy.js:4]: it should be in configs

File: cypress/e2e/test.cy.js [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  Cypress.on("uncaught:exception", (err, runnable) => {
```
Comment (nedo) [inline cypress/e2e/test.cy.js:5]: Please explain reason of having this code?

File: cypress/e2e/test.cy.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+    cy.get("#salary").type(12345);
+    cy.get("#department").type("QA");
+    cy.get("#submit").click();
+
+    cy.get(".rt-tr-group:eq(3) .rt-td:eq(0)").should("contain", "Alden");
+    cy.get(".rt-tr-group:eq(3) .rt-td:eq(1)").should("contain", "Cantrell");
+    cy.get(".rt-tr-group:eq(3) .rt-td:eq(2)").should("contain", "32");
+    cy.get(".rt-tr-group:eq(3) .rt-td:eq(3)").should(
+      "contain",
+      "test@test.com"
```
Comment (nedo) [inline cypress/e2e/test.cy.js:33]: Separate Test Data

**Comment by nedo** [comment]
1.  Please Restructure code following Page Object Model technique,  
- https://www.browserstack.com/guide/cypress-page-object-model
- https://medium.com/nerd-for-tech/cypress-page-object-model-953791736972
2. Please write/convert your code in Typescript 
3. Almost all selectors are weak, please select strong CSS selectors
4. Add ReadMi how can I trigger these test cases? 

@Kashif-Rehman please fix/answer points and tag me, I'll reivew it again. Please let me know if there are any questions. Thanks.

**Review by nedo — CHANGES_REQUESTED** [review]

**Comment by nedo** [comment]
Please start reading following two roadmaps 

- Automation - https://roadmap.sh/typescript 
- Manual Testing - https://roadmap.sh/qa

Let me know if some help is needed. Thanks

File: cypress/e2e/test.cy.js (hunk 2 above)
Comment (Kashif-Rehman) [reply cypress/e2e/test.cy.js]: Line 53 and 54 are updated with POM implementation 
for uploading image uploading

File: cypress/e2e/API.cy.js (hunk 3 above)
Comment (Kashif-Rehman) [reply cypress/e2e/API.cy.js:18]: Forgot to push the fixture file.

File: cypress/e2e/API.cy.js (hunk 4 above)
Comment (Kashif-Rehman) [reply cypress/e2e/API.cy.js:33]: Forgot to push the fixture file.
Bearer token required here

File: cypress/e2e/test.cy.js (hunk 5 above)
Comment (Kashif-Rehman) [reply cypress/e2e/test.cy.js:10]: done

File: cypress/e2e/test.cy.js (hunk 6 above)
Comment (Kashif-Rehman) [reply cypress/e2e/test.cy.js:19]: Line No 1 for implementing drag and drop 
Line No 5. For controlling errors other thin cypress

File: cypress/e2e/test.cy.js (hunk 7 above)
Comment (Kashif-Rehman) [reply cypress/e2e/test.cy.js]: done

**Comment by nedo** [comment]
@Kashif-Rehman please complete some pending & more points, 
- Please write/convert your code in Typescript
- In POM there should be BasePage having common page contents - which should be reused
- Implement drag-and-drop without plugin, use Cypress document for that 
- Up next start exploring parameterization and use-case would be to fill form( `it("Verify user can enter new data into table", ())` with multiple rows 
- explore and add script in package.json

**Comment by Kashif-Rehman** [comment]
- Please write/convert your code in Typescript (done)
- In POM there should be BasePage having common page contents - which should be reused (done)
- Implement drag-and-drop without plugin, use Cypress document for that (done)
- Up next start exploring parameterization and use-case would be to fill form( it("Verify user can enter new data into table", ()) with multiple rows (done)

Please explain the "explore and add script in package.json"

please help me the

**Comment by nedo** [comment]
Pending points, 
- Implement drag-and-drop without plugin, use Cypress document for that - please follow instructions 
- In POM there should be BasePage having common page contents - which should be reused > In BasePage please put more contents i.e click(), type() it should return something which should be asserted at test case level 

Further Tasks 
- add gitignore 
- integrate EsLinting 
- integrate cypress dashboard 
- start exploring `interfaces` & `abstraction` 
Let me know once done - I'll start triggering cases at my machine and move futher. Thanks

**Review by nedo — CHANGES_REQUESTED** [review]

File: cypress/pages/interactions.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
+import BasePage from "./basePage";
+import "@4tw/cypress-drag-drop";
```
[the remark this answers is not in the document]
Comment (Kashif-Rehman) [reply cypress/pages/interactions.ts:2]: Drag and drop is implemented with both methods with plugin and without plugin 
 it("Verify user can drag and drop", () => {
    home.clickInteractions();
    interactions.clickDroppable();
    cy.wait(5000);
    interactions.dragDrop();
  });

  it("Drag and drop without library", () => {
    home.clickInteractions();
    interactions.clickDroppable();
    interactions.dragWithoutLibrary();
  });

**Comment by nedo** [comment]
@Kashif-Rehman added some more comments, to keep things simple please merge this PR and handle further changes into next PR. 

Points for next PR 
1. Create an element interface having definition of all needed functions 
2. Create elements class(implementing above interface) having implementations of all above functions
3. All pages should use these functions further

**Pull request merged**